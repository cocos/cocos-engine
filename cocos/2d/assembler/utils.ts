/*
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { Color, Mat4, clamp } from '../../core';
import { RenderData } from '../renderer/render-data';
import { IBatcher } from '../renderer/i-batcher';
import { Node } from '../../scene-graph/node';
import { FormatInfos } from '../../gfx';

const m = new Mat4();

export function fillMeshVertices3D (node: Node, renderer: IBatcher, renderData: RenderData, color: Color) {
    const chunk = renderData.chunk;
    const dataList = renderData.data;
    const vData = chunk.vb;
    const vertexCount = renderData.vertexCount;

    node.getWorldMatrix(m);

    let vertexOffset = 0;
    for (let i = 0; i < vertexCount; i++) {
        const vert = dataList[i];
        const x = vert.x;
        const y = vert.y;
        let rhw = m.m03 * x + m.m07 * y + m.m15;
        rhw = rhw ? Math.abs(1 / rhw) : 1;
        vData[vertexOffset + 0] = (m.m00 * x + m.m04 * y + m.m12) * rhw;
        vData[vertexOffset + 1] = (m.m01 * x + m.m05 * y + m.m13) * rhw;
        vData[vertexOffset + 2] = (m.m02 * x + m.m06 * y + m.m14) * rhw;
        Color.toArray(vData, color, vertexOffset + 5);
        vertexOffset += 9;
    }

    // fill index data
    const bid = chunk.bufferId;
    const vid = chunk.vertexOffset;
    const meshBuffer = chunk.meshBuffer;
    const ib = chunk.meshBuffer.iData;
    let indexOffset = meshBuffer.indexOffset;
    for (let i = 0, count = vertexCount / 4; i < count; i++) {
        const start = vid + i * 4;
        ib[indexOffset++] = start;
        ib[indexOffset++] = start + 1;
        ib[indexOffset++] = start + 2;
        ib[indexOffset++] = start + 1;
        ib[indexOffset++] = start + 3;
        ib[indexOffset++] = start + 2;
    }
    meshBuffer.indexOffset += renderData.indexCount;
    meshBuffer.setDirty();
}

export function updateOpacity (renderData: RenderData, opacity: number) {
    const vfmt = renderData.vertexFormat;
    const vb = renderData.chunk.vb;
    let attr; let format; let stride;
    // Color component offset
    let offset = 0;
    for (let i = 0; i < vfmt.length; ++i) {
        attr = vfmt[i];
        format = FormatInfos[attr.format];
        if (format.hasAlpha) {
            stride = renderData.floatStride;
            if (format.size / format.count === 1) {
                const alpha = ~~clamp(Math.round(opacity * 255), 0, 255);
                // Uint color RGBA8
                for (let color = offset; color < vb.length; color += stride) {
                    vb[color] = ((vb[color] & 0xffffff00) | alpha) >>> 0;
                }
            } else if (format.size / format.count === 4) {
                // RGBA32 color, alpha at position 3
                for (let alpha = offset + 3; alpha < vb.length; alpha += stride) {
                    vb[alpha] = opacity;
                }
            }
        }
        offset += format.size >> 2;
    }
}
