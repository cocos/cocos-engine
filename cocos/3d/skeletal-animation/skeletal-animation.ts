/*
 Copyright (c) 2017-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import {
    ccclass, executeInEditMode, executionOrder, help, menu, tooltip, type, serializable, editable,
} from 'cc.decorator';
import { SkinnedMeshRenderer } from '../skinned-mesh-renderer';
import { Mat4, cclegacy, js, assertIsTrue } from '../../core';
import { DataPoolManager } from './data-pool-manager';
import { Node } from '../../scene-graph/node';
import { AnimationClip } from '../../animation/animation-clip';
import { Animation } from '../../animation/animation-component';
import { SkelAnimDataHub } from './skeletal-animation-data-hub';
import { SkeletalAnimationState } from './skeletal-animation-state';
import { getWorldTransformUntilRoot } from '../../animation/transform-utils';
import type { AnimationState } from '../../animation/animation-state';
import { getGlobalAnimationManager } from '../../animation/global-animation-manager';

/**
 * @en The socket to synchronize transform from skeletal joint to target node.
 * @zh 骨骼动画的挂点，用于将骨骼树的挂点节点变化矩阵同步到目标节点上
 */
@ccclass('cc.SkeletalAnimation.Socket')
export class Socket {
    /**
     * @en Path of the target joint.
     * @zh 此挂点的目标骨骼路径。
     */
    @serializable
    @editable
    public path = '';

    /**
     * @en Transform output node.
     * @zh 此挂点的变换信息输出节点。
     */
    @type(Node)
    public target: Node | null = null;

    constructor (path = '', target: Node | null = null) {
        this.path = path;
        this.target = target;
    }
}

js.setClassAlias(Socket, 'cc.SkeletalAnimationComponent.Socket');

const m4_1 = new Mat4();
const m4_2 = new Mat4();

function collectRecursively (node: Node, prefix = '', out: string[] = []) {
    for (let i = 0; i < node.children.length; i++) {
        const child = node.children[i];
        if (!child) { continue; }
        const path = prefix ? `${prefix}/${child.name}` : child.name;
        out.push(path);
        collectRecursively(child, path, out);
    }
    return out;
}

/**
 * @en
 * Skeletal animation component, offers the following features on top of [[Animation]]:
 * * Choice between baked animation and real-time calculation, to leverage efficiency and expressiveness.
 * * Joint socket system: Create any socket node directly under the animation component root node,
 *   find your target joint and register both to the socket list, so that the socket node would be in-sync with the joint.
 * @zh
 * 骨骼动画组件，在普通动画组件基础上额外提供以下功能：
 * * 可选预烘焙动画模式或实时计算模式，用以权衡运行时效率与效果；
 * * 提供骨骼挂点功能：通过在动画根节点下创建挂点节点，并在骨骼动画组件上配置 socket 列表，挂点节点的 Transform 就能与骨骼保持同步。
 */
@ccclass('cc.SkeletalAnimation')
@help('i18n:cc.SkeletalAnimation')
@executionOrder(99)
@executeInEditMode
@menu('Animation/SkeletalAnimation')
export class SkeletalAnimation extends Animation {
    public static Socket = Socket;

    /**
     * @en
     * The joint sockets this animation component maintains.<br>
     * Sockets have to be registered here before attaching custom nodes to animated joints.
     * @zh
     * 当前动画组件维护的挂点数组。要挂载自定义节点到受动画驱动的骨骼上，必须先在此注册挂点。
     */
    @type([Socket])
    @tooltip('i18n:animation.sockets')
    get sockets () {
        return this._sockets;
    }

    set sockets (val) {
        if (!this._useBakedAnimation) {
            const animMgr = getGlobalAnimationManager();
            animMgr.removeSockets(this.node, this._sockets);
            animMgr.addSockets(this.node, val);
        }
        this._sockets = val;
        this.rebuildSocketAnimations();
    }

    /**
     * @en
     * Whether to bake animations. Default to true,<br>
     * which substantially increases performance while making all animations completely fixed.<br>
     * Dynamically changing this property will take effect when playing the next animation clip.
     * @zh
     * 是否使用预烘焙动画，默认启用，可以大幅提高运行效时率，但所有动画效果会被彻底固定，不支持任何形式的编辑和混合。<br>
     * 运行时动态修改此选项会在播放下一条动画片段时生效。
     */
    @tooltip('i18n:animation.use_baked_animation')
    get useBakedAnimation () {
        return this._useBakedAnimation;
    }

    set useBakedAnimation (val) {
        this._useBakedAnimation = val;

        for (const stateName in this._nameToState) {
            const state = this._nameToState[stateName] as SkeletalAnimationState;
            state.setUseBaked(val);
        }

        this._users.forEach((user) => {
            user.setUseBakedAnimation(val);
        });

        if (this._useBakedAnimation) {
            getGlobalAnimationManager().removeSockets(this.node, this._sockets);
        } else {
            getGlobalAnimationManager().addSockets(this.node, this._sockets);
            this._currentBakedState = null;
        }
    }

    @serializable
    protected _useBakedAnimation = true;

    @type([Socket])
    protected _sockets: Socket[] = [];

    public onLoad () {
        super.onLoad();
        // Actively search for potential users and notify them that an animation is usable.
        const comps = this.node.getComponentsInChildren(SkinnedMeshRenderer);
        for (let i = 0; i < comps.length; ++i) {
            const comp = comps[i];
            if (comp.skinningRoot === this.node) {
                this.notifySkinnedMeshAdded(comp);
            }
        }
    }

    public onDestroy () {
        super.onDestroy();
        (cclegacy.director.root.dataPoolManager as DataPoolManager).jointAnimationInfo.destroy(this.node.uuid);
        getGlobalAnimationManager().removeSockets(this.node, this._sockets);
        this._removeAllUsers();
    }

    public onEnable () {
        super.onEnable();
        this._currentBakedState?.resume();
    }

    public onDisable () {
        super.onDisable();
        this._currentBakedState?.pause();
    }

    public start () {
        this.sockets = this._sockets;
        this.useBakedAnimation = this._useBakedAnimation;
        super.start();
    }

    public pause () {
        if (!this._useBakedAnimation) {
            super.pause();
        } else {
            this._currentBakedState?.pause();
        }
    }

    public resume () {
        if (!this._useBakedAnimation) {
            super.resume();
        } else {
            this._currentBakedState?.resume();
        }
    }

    public stop () {
        if (!this._useBakedAnimation) {
            super.stop();
        } else if (this._currentBakedState) {
            this._currentBakedState.stop();
            this._currentBakedState = null;
        }
    }

    /**
     * @en Query all socket paths
     * @zh 获取所有挂点的骨骼路径
     * @returns @en All socket paths @zh 所有挂点的骨骼路径
     */
    public querySockets () {
        const animPaths = (this._defaultClip && Object.keys(SkelAnimDataHub.getOrExtract(this._defaultClip).joints).sort()
            .reduce((acc, cur) => (cur.startsWith(`${acc[acc.length - 1]}/`) ? acc : (acc.push(cur), acc)), [] as string[])) || [];
        if (!animPaths.length) { return ['please specify a valid default animation clip first']; }
        const out: string[] = [];
        for (let i = 0; i < animPaths.length; i++) {
            const path = animPaths[i];
            const node = this.node.getChildByPath(path);
            if (!node) { continue; }
            out.push(path);
            collectRecursively(node, path, out);
        }
        return out;
    }

    /**
     * @en Rebuild animations to synchronize immediately all sockets to their target node.
     * @zh 重建动画并立即同步所有挂点的转换矩阵到它们的目标节点上。
     */
    public rebuildSocketAnimations () {
        for (const socket of this._sockets) {
            const joint = this.node.getChildByPath(socket.path);
            const { target } = socket;
            if (joint && target) {
                target.name = `${socket.path.substring(socket.path.lastIndexOf('/') + 1)} Socket`;
                target.parent = this.node;
                getWorldTransformUntilRoot(joint, this.node, m4_1);
                Mat4.fromRTS(m4_2, target.rotation, target.position, target.scale);
                if (!Mat4.equals(m4_2, m4_1)) { target.matrix = m4_1; }
            }
        }
        for (const stateName of Object.keys(this._nameToState)) {
            const state = this._nameToState[stateName] as SkeletalAnimationState;
            state.rebuildSocketCurves(this._sockets);
        }
    }

    /**
     * @en Create or get the target node from a socket.
     * If a socket haven't been created for the corresponding path, this function will register a new socket.
     * @zh 创建或获取一个挂点的同步目标节点。
     * 如果对应路径还没有创建挂点，这个函数会创建一个新的挂点。
     * @param path @en Path of the target joint. @zh 此挂点的骨骼路径。
     * @returns @en The target node of the socket. @zh 挂点的目标节点
     */
    public createSocket (path: string) {
        const socket = this._sockets.find((s) => s.path === path);
        if (socket) { return socket.target; }
        const joint = this.node.getChildByPath(path);
        if (!joint) { console.warn('illegal socket path'); return null; }
        const target = new Node();
        target.parent = this.node;
        this._sockets.push(new Socket(path, target));
        this.rebuildSocketAnimations();
        return target;
    }

    /**
     * @internal This method only friends to skinned mesh renderer.
     */
    public notifySkinnedMeshAdded (skinnedMeshRenderer: SkinnedMeshRenderer) {
        const { _useBakedAnimation: useBakedAnimation } = this;
        const formerBound = skinnedMeshRenderer.associatedAnimation;
        if (formerBound) {
            formerBound._users.delete(skinnedMeshRenderer);
        }
        skinnedMeshRenderer.associatedAnimation = this;
        skinnedMeshRenderer.setUseBakedAnimation(useBakedAnimation, true);
        if (useBakedAnimation) {
            const { _currentBakedState: playingState } = this;
            if (playingState) {
                skinnedMeshRenderer.uploadAnimation(playingState.clip);
            }
        }
        this._users.add(skinnedMeshRenderer);
    }

    /**
     * @internal This method only friends to skinned mesh renderer.
     */
    public notifySkinnedMeshRemoved (skinnedMeshRenderer: SkinnedMeshRenderer) {
        assertIsTrue(skinnedMeshRenderer.associatedAnimation === this || skinnedMeshRenderer.associatedAnimation === null);
        skinnedMeshRenderer.setUseBakedAnimation(false);
        skinnedMeshRenderer.associatedAnimation = null;
        this._users.delete(skinnedMeshRenderer);
    }

    /**
     * Get all users.
     * @internal This method only friends to the skeleton animation state.
     */
    public getUsers () {
        return this._users;
    }

    protected _createState (clip: AnimationClip, name?: string) {
        return new SkeletalAnimationState(clip, name);
    }

    protected _doCreateState (clip: AnimationClip, name: string) {
        const state = super._doCreateState(clip, name) as SkeletalAnimationState;
        state.rebuildSocketCurves(this._sockets);
        return state;
    }

    protected doPlayOrCrossFade (state: AnimationState, duration: number) {
        if (this._useBakedAnimation) {
            if (this._currentBakedState) {
                this._currentBakedState.stop();
            }
            const skeletalAnimationState = state as SkeletalAnimationState;
            this._currentBakedState = skeletalAnimationState;
            skeletalAnimationState.play();
        } else {
            super.doPlayOrCrossFade(state, duration);
        }
    }

    private _users = new Set<SkinnedMeshRenderer>();

    private _currentBakedState: SkeletalAnimationState | null = null;

    private _removeAllUsers () {
        Array.from(this._users).forEach((user) => {
            this.notifySkinnedMeshRemoved(user);
        });
    }
}
