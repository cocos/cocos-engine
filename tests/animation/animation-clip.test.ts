import { Color, js, lerp, Quat, Rect, Size, Vec2, Vec3, Vec4 } from '../../cocos/core';
import { AnimationClip, AnimationState, AnimationManager } from '../../cocos/animation';
import { AnimationController, ColorTrack, ComponentPath, HierarchyPath, IValueProxyFactory, RealTrack, SizeTrack, Track, VectorTrack } from '../../cocos/animation/animation';
import { ccclass } from 'cc.decorator';
import { captureErrorIDs, captureWarnIDs } from '../utils/log-capture';
import { Node,Component } from '../../cocos/scene-graph';
import { LegacyBlendStateBuffer } from '../../cocos/3d/skeletal-animation/skeletal-animation-blending';
import { AnimationGraph } from '../../cocos/animation/marionette/animation-graph';
import { ClipMotion } from '../../cocos/animation/marionette/clip-motion';
import { AnimationGraphEval } from '../../cocos/animation/marionette/graph-eval';

test('Common target', () => {
    @ccclass('TestComponent')
    class TestComponent extends Component {
        get value () {
            return this.c;
        }

        set value (value: Vec3) {
            Vec3.copy(this.c, value);
        }

        public c: Vec3 = new Vec3(0, -1, 0);
    }

    const node = new Node();

    const testComponent = node.addComponent(TestComponent);
    const animationClip = new AnimationClip();
    animationClip.wrapMode = AnimationClip.WrapMode.Loop;
    animationClip.duration = 2;
    animationClip.keys = [
        [0, 1, 2],
    ];
    animationClip.commonTargets = [{
        modifiers: [
            new ComponentPath(js.getClassName(TestComponent)),
            'value',
        ],
    }];
    animationClip.curves = [
        {
            commonTarget: 0,
            modifiers: [ 'x' ],
            data: {
                keys: 0,
                values: [ 0, 0.5, 1 ],
            }
        },
        {
            commonTarget: 0,
            modifiers: [ 'z' ],
            data: {
                keys: 0,
                values: [ 0, 0.5, 1 ],
            }
        },
    ];
    
    const expects: Record<number, Vec3> = {
        0: new Vec3(0, -1, 0),
        1: new Vec3(0.5, -1, 0.5),
        2: new Vec3(1, -1, 1),
    };

    const animationState = new AnimationState(animationClip);
    animationState.initialize(node);
    for (const time of Object.keys(expects)) {
        const c = expects[time];
        animationState.setTime(parseInt(time));
        animationState.sample();
        expect(testComponent.c).toEqual(c);
    }
});

test('Common targets that modify eulerAngles', () => {
    const animationManager = new AnimationManager();
    const mockInstance = jest.spyOn((global as any).cc.director, 'getAnimationManager').mockImplementation(() => {
        return animationManager;
    });

    const partialNodeName = 'partial';

    const clip = new AnimationClip('default');
    clip.duration = 2.0;
    const keys = [0, 0.1, 0.2];
    const eulerValues = [new Vec3(), new Vec3(1), new Vec3(2)];
    const eulerXValues = [0, 1, 2];
    clip.keys = [keys];
    clip.commonTargets = [{
        modifiers: [new HierarchyPath(partialNodeName), 'eulerAngles'],
    }];
    clip.curves = [{
        modifiers: ['eulerAngles'],
        data: {keys: 0, values: eulerValues},
    }, {
        commonTarget: 0, // euler angles
        modifiers: ['x'],
        data: {keys: 0, values: eulerXValues},
    }];

    const state = new AnimationState(clip, clip.name);
    state.weight = 1.0;
    
    const node = new Node();
    const partialNode = new Node(partialNodeName);
    node.addChild(partialNode);
    state.initialize(node);

    animationManager.update(0);
    state.play();
    state.pause();

    state.setTime(keys[1]);
    state.update(0);
    animationManager.update(0);
    
    expect(node.eulerAngles).toEqual(eulerValues[1]);
    expect(partialNode.eulerAngles.x).toEqual(eulerXValues[1]);

    mockInstance.mockRestore();
});

describe('Custom track setter', () => {
    class Target {
        public setValue(value: Target['_value']) { Vec3.copy(this._value, value); }
        public getValue() { return this._value; }
        private _value = { x: 0, y: 0, z: 0 };
    }

    const valueProxyWithOnlySet: IValueProxyFactory = {
        forTarget: (target: Target) => {
            return {
                set: (value: Target['_value']) => { target.setValue(value) },
            };
        },
    };

    const valueProxyWithGetSet: IValueProxyFactory = {
        forTarget: (target: Target) => {
            return {
                set: (value: Target['_value']) => { target.setValue(value) },
                get: () => target.getValue(),
            };
        },
    };

    test('get() got not called if non of channels is empty', () => {
        const target = new Target();
        const mockGetValue = target.getValue = jest.fn(target.getValue);
        const mockSetValue = target.setValue = jest.fn(target.setValue);
    
        const track = new VectorTrack();
        track.proxy = valueProxyWithGetSet;
        track.channels().forEach(({ curve }) => {
            curve.assignSorted([[0.0, ({ value: 0.0 })]]);
        });

        const clip = new AnimationClip();
        clip.addTrack(track);
        const clipEval = clip.createEvaluator({
            target,
        });
        clipEval.evaluate(0.0);
        expect(mockGetValue).not.toBeCalled();
        expect(mockSetValue).toBeCalled();
    });

    test('If get() is not defined, the default channel value would be used', () => {
        const target = new Target();
        const mockSetValue = target.setValue = jest.fn(target.setValue);
    
        const track = new VectorTrack();
        track.proxy = valueProxyWithOnlySet;
        track.channels().forEach(({ curve }) => {
            curve.assignSorted([[0.0, ({ value: 2.0 })]]);
        });
        const clip = new AnimationClip();
        clip.addTrack(track);
        const clipEval = clip.createEvaluator({
            target,
        });
        clipEval.evaluate(0.0);
        expect(mockSetValue).toBeCalled();
    });
});

test('animation state', () => {
    const animationManager = new AnimationManager();
    const mockInstance = jest.spyOn((global as any).cc.director, 'getAnimationManager').mockImplementation(() => {
        return animationManager;
    });

    const clip = new AnimationClip('default');
    clip.duration = 2.0;
    const keys = [
        0,
        0.1,
        0.2,
    ];
    const values = [
        new Vec3(),
        new Vec3(1),
        new Vec3(2),
    ];
    clip.keys = [keys];
    clip.curves = [{
        modifiers: [
            'position',
        ],
        data: {
            keys: 0,
            values,
        },
    }];

    const state = new AnimationState(clip, clip.name);
    state.weight = 1.0;
    
    const node = new Node();
    state.initialize(node);

    animationManager.update(0);
    state.play();
    state.pause();

    state.setTime(keys[1]);
    state.update(0);
    animationManager.update(0);
    expect(node.getPosition()).toEqual(values[1]);

    mockInstance.mockRestore();
});

test('default animation clip validation', () => {
    const validClip = new AnimationClip('valid');
    expect(validClip.validate()).toEqual(true);
});

test('Warn on track binding failure', () => {
    const clip = new AnimationClip();
    clip.name = 'AnyClip';
    const track = new RealTrack();
    track.path.toHierarchy('meow');
    // Ensure the track is not empty otherwise its instantiation will be skipped.
    track.channel.curve.assignSorted([0.0], [1.0]);
    clip.addTrack(track);
    const node = new Node('AnyRoot');

    const warnIDWatcher = captureWarnIDs();
    const errorIDWatcher = captureErrorIDs(); // To silence the hierarchy binding error which we don't care.
    void clip.createEvaluator({
        target: node,
    });
    expect(errorIDWatcher.captured.length).toBe(1); // No further verification -- it doesn't concern.
    errorIDWatcher.stop();
    expect(warnIDWatcher.captured.length).toBeGreaterThan(0);
    const [lastId, ...lastArgs] = warnIDWatcher.captured[warnIDWatcher.captured.length - 1];
    expect(lastId).toBe(3937);
    expect(lastArgs).toStrictEqual([
        clip.name,
        node.name,
    ]);
    warnIDWatcher.stop();
});

describe(`Component-wise animation`, () => {
    /**
     * This test the component-wise animation capability of animation clips.
     * Both traditional environment(animation state) and marionette environment(animation graph) are tested.
     * 
     * Our test goal includes `Vec2`, `Vec3`, `Vec4`, `Color`, `Size` objects.
     * For animation graph environment, we only tested `Vec3` since animation graph only interests on `Vec3`.
     * Other types of track have no such component-wise animation capabilities.
     */
    const _note = undefined;

    const newVecXTrack = (x: 2 | 3 | 4) => { const track = new VectorTrack(); track.componentsCount = x; return track; };

    /** The specs for each value type. See `Spec` for description. */
    const SPECS = [
        { // vec2
            constructor: Vec2,
            defaultKeys: [['x', 0]] as const,
            animatedKeys: [['y', 1]] as const,
            createTrack: () => newVecXTrack(2),
        } as Spec<Vec2>,
        { // vec3
            constructor: Vec3,
            defaultKeys: [['y', 1]] as const,
            animatedKeys: [['x', 0], ['z', 2]] as const,
            createTrack: () => newVecXTrack(3),
        } as Spec<Vec3>,
        { // vec4
            constructor: Vec4,
            defaultKeys: [['y', 1], ['z', 2]] as const,
            animatedKeys: [['x', 0], ['w', 3]] as const,
            createTrack: () => newVecXTrack(4),
        } as Spec<Vec4>,
        { // color
            constructor: Color,
            range: [0, 255],
            defaultKeys: [['g', 1], ['b', 2]] as const,
            animatedKeys: [['r', 0], ['a', 3]] as const,
            createTrack: () => new ColorTrack(),
        } as Spec<Color>,
        { // size
            constructor: Size,
            defaultKeys: [['height', 1]] as const,
            animatedKeys: [['width', 0]] as const,
            createTrack: () => new SizeTrack()
        } as Spec<Size>,
    ];

    test(`Traditional`, () => {
        check(evaluateClipSingleFrameTraditional, SPECS);
    });

    test(`Animation graph`, () => {
        // For animation graph we only check vec3 since currently we don't want to support other type of value types.
        check(
            evaluateClipSingleFrameMarionette,
            SPECS.filter((spec) => spec.constructor === Vec3),
        );
    });

    function check(evaluate: EvaluateClipSingleFrame, specs: typeof SPECS) {
        const node = new Node();
        
        @ccclass('DummyComponent')
        class DummyComponent extends Component { }
        const component = node.addComponent(DummyComponent);

        const clip = new AnimationClip();
        clip.duration = 1.0;

        const TIME = 0.2;

        // Generate test data.
        const expects = specs.map(({ constructor, defaultKeys, animatedKeys, range, createTrack }) => {
            /** Generates a fixed value at [tMin, tMax) according to array index and array length. */
            const gen = (keyIndex: number, keyCount: number, tMin: number, tMax: number) => {
                const t = tMin + (tMax - tMin) * (keyIndex / keyCount);
                return range ? lerp(range[0], range[1], t) : t;
            };

            // Generate default value for each key at [0.1, 0.5).
            const defaultValue = new constructor();
            const defaultProperties: Record<string, number> = {};
            [...defaultKeys, ...animatedKeys].sort().forEach(([propertyKey, _], keyIndex, keys) => {
                const propertyValue = gen(keyIndex, keys.length, 0.1, 0.5);
                defaultValue[propertyKey] = propertyValue;
                defaultProperties[propertyKey] = defaultValue[propertyKey]; // Don't use `= propertyValue` here since for example, color matters
            });

            // Register the value to the dummy component so animation system can animate it.
            if (constructor === Vec3) { // Specially handle `Vec3`
                node.position = (defaultValue as Vec3).clone();
            } else {
                Object.defineProperty(component, constructor.name, { value: defaultValue.clone(), writable: true });
            }

            // Create the track and generate some animation values at [0.5, 1.0) for animated keys.
            const track = createTrack();
            if (constructor === Vec3) {
                track.path.toProperty('position');
            } else {
                track.path.toComponent(DummyComponent).toProperty(constructor.name);
            }
            const expectedAnimatedProperties: Record<string, number> = {};
            [...animatedKeys].forEach(([propertyKey, channelIndex], keyIndex, keys) => {
                const propertyAnimationValue = gen(keyIndex, keys.length, 0.5, 1.0);
                [...track.channels()][channelIndex].curve.addKeyFrame(0.0, propertyAnimationValue);
                const o = new constructor();
                o[propertyKey] = propertyAnimationValue;
                expectedAnimatedProperties[propertyKey] = o[propertyKey];
            });
            clip.addTrack(track);

            return {
                specName: constructor.name, // To showup in snapshot
                defaultProperties,
                expectedAnimatedProperties,
            };
        });

        // Let's inline the expectation so we can check manually 👀.
        expect(expects).toMatchSnapshot();

        // Evaluate.
        evaluate(clip, node, TIME);

        // Check test data.
        specs.forEach(({ constructor, defaultKeys, animatedKeys }, specIndex) => {
            const {
                defaultProperties,
                expectedAnimatedProperties,
            } = expects[specIndex];
            const object = constructor === Vec3
                ? node.position
                : component[constructor.name];
            // These properties should keep default.
            for (const [propertyKey] of defaultKeys) {
                expect(object[propertyKey]).toBe(defaultProperties[propertyKey]);
            }
            // These properties should be animated.
            for (const [propertyKey] of animatedKeys) {
                expect(object[propertyKey]).toBeCloseTo(expectedAnimatedProperties[propertyKey]);
            }
        });

        js.unregisterClass(DummyComponent);
    }

    type PropertyKeyAndChannelIndex<T> = [
        /** The property key we're going to observe. */
        propertyKey: keyof T,

        /** The corresponding channel index in created track. */
        channelIndex: number,
    ];

    /** Describes a spec to test "component-wise animation". */
    interface Spec<T extends { clone(): T }> {
        /** The value's constructor. */
        constructor: new () => T;

        /** These keys of the value will be checked to remain default. */
        defaultKeys: readonly Readonly<PropertyKeyAndChannelIndex<T>>[];

        /** These keys of the value will be checked to be corresponding animation values. */
        animatedKeys: readonly Readonly<PropertyKeyAndChannelIndex<T>>[];

        /** Creates a track which yields such a value. */
        createTrack: () => Track;

        /** By default, the test routine generates values at (0, 1). Use this field to map the value into specified range. */
        range?: [min: number, max:number];
    }
});

type EvaluateClipSingleFrame = (clip: AnimationClip, node: Node, time: number) => void;

function evaluateClipSingleFrameTraditional(clip: AnimationClip, node: Node, time: number) {
    const blendStateBuffer = new LegacyBlendStateBuffer();
    const animationState = new AnimationState(clip);
    animationState.initialize(node, blendStateBuffer);
    animationState.time = time;
    animationState.sample();
}

function evaluateClipSingleFrameMarionette (clip: AnimationClip, node: Node, time: number) {
    const animationGraph = new AnimationGraph();
    const layer = animationGraph.addLayer();
    const motion = layer.stateMachine.addMotion();
    const clipMotion = motion.motion = new ClipMotion();
    clipMotion.clip = clip;
    layer.stateMachine.connect(layer.stateMachine.entryState, motion);
    const controller = node.addComponent(AnimationController) as AnimationController;
    const graphEval = new AnimationGraphEval(animationGraph, node, controller);
    graphEval.update(time);
}