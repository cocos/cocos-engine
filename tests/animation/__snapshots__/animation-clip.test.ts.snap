// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component-wise animation Animation graph 1`] = `
Array [
  Object {
    "defaultProperties": Object {
      "x": 0.1,
      "y": 0.23333333333333334,
      "z": 0.3666666666666667,
    },
    "expectedAnimatedProperties": Object {
      "x": 0.5,
      "z": 0.75,
    },
    "specName": "Vec3",
  },
]
`;

exports[`Component-wise animation Traditional 1`] = `
Array [
  Object {
    "defaultProperties": Object {
      "x": 0.1,
      "y": 0.30000000000000004,
    },
    "expectedAnimatedProperties": Object {
      "y": 0.5,
    },
    "specName": "Vec2",
  },
  Object {
    "defaultProperties": Object {
      "x": 0.1,
      "y": 0.23333333333333334,
      "z": 0.3666666666666667,
    },
    "expectedAnimatedProperties": Object {
      "x": 0.5,
      "z": 0.75,
    },
    "specName": "Vec3",
  },
  Object {
    "defaultProperties": Object {
      "w": 0.1,
      "x": 0.2,
      "y": 0.30000000000000004,
      "z": 0.4,
    },
    "expectedAnimatedProperties": Object {
      "w": 0.75,
      "x": 0.5,
    },
    "specName": "Vec4",
  },
  Object {
    "defaultProperties": Object {
      "a": 25,
      "b": 51,
      "g": 76,
      "r": 102,
    },
    "expectedAnimatedProperties": Object {
      "a": 191,
      "r": 127,
    },
    "specName": "Color",
  },
  Object {
    "defaultProperties": Object {
      "height": 0.1,
      "width": 0.30000000000000004,
    },
    "expectedAnimatedProperties": Object {
      "width": 0.5,
    },
    "specName": "Size",
  },
]
`;
